File: TFG_SIGA/SIGA/src/main/java/com/siga/entity/Boleta.java
package com.siga.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;

@Entity
@Getter
@Setter
public class Boleta {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long idBoleta;
    
    @ManyToOne
    @JoinColumn(name = "usuario_id")
    private User usuario;

    @NotNull
    private String nombreProducto;

    private String tienda;
    private Double precio;
    private LocalDate fechaCompra;
    private String imagenPath;

    @Column(name = "textoocr", columnDefinition = "TEXT")
    private String textoOCR;

    @ManyToOne
    @JoinColumn(name = "tipo_producto_id")
    private TipoProducto tipoProducto;

    @OneToOne(mappedBy = "boleta", cascade = CascadeType.ALL, orphanRemoval = true)
    private Garantia garantia;

}

File: TFG_SIGA/siga_app/lib/models/boleta.dart
class Boleta {

  final int idBoleta;
  final String nombreProducto;
  final String? tienda;
  final double? precio;
  final DateTime fechaCompra;
  final String? imagenPath;
  final String? textoOCR;
  final int diasRestantesGarantia;

  final String? categoria;
  final DateTime? fechaFinGarantia;

  Boleta({
    required this.idBoleta,
    required this.nombreProducto,
    required this.fechaCompra,
    required this.diasRestantesGarantia,

    this.tienda,
    this.precio,
    this.imagenPath,
    this.textoOCR,

    this.categoria,
    this.fechaFinGarantia,
  });

  factory Boleta.fromJson(Map<String, dynamic> json) {

    return Boleta(

      idBoleta: json['idBoleta'],

      nombreProducto: json['nombreProducto'],

      tienda: json['tienda'],

      precio: json['precio'] != null
          ? (json['precio'] as num).toDouble()
          : null,

      fechaCompra: DateTime.parse(json['fechaCompra']),

      imagenPath: json['imagenPath'],

      textoOCR: json['textoOCR'],

      diasRestantesGarantia:
          json['diasRestantesGarantia'] ?? 0,

      categoria: json['categoria'],

      fechaFinGarantia:
          json['fechaFinGarantia'] != null
              ? DateTime.parse(json['fechaFinGarantia'])
              : null,
    );
  }
}